kaitaistruct.py

```diff
@@ -293,6 +385,45 @@ def read_bytes(self, n):
             )
         return r
 
+    def write_bytes(self, data):
+        if data is None:
+            return
+        nb = len(data)
+        if nb > 0 and self._io.write(data) != nb:
+            raise ValidationFailedError(
+                "not all of %d bytes written" %
+                (nb,), self._io, __file__
+            )
```

Comment:
1. The class `ValidationFailedError` is reserved for custom explicit format validations inside the `.ksy` spec (using the `valid` key, see https://github.com/kaitai-io/kaitai_struct/issues/435#issue-323332353), not for internal erroneous states of the runtime library. The point is that if the application code catches exception of type `ValidationFailedError`, the programmer can be sure that it's caused by a check performed by `contents` or `valid` key that he defined himself in the KSY format definition.

2. The `self._io` accessed inside the `KaitaiStream` class yields the original [`io.IOBase`](https://docs.python.org/3/library/io.html#io.IOBase)-derived stream resource, not the expected `KaitaiStream` object.

   Initializing will throw `AttributeError: '_io.BytesIO' object has no attribute 'pos'`, because the [`BytesIO` object](https://docs.python.org/3/library/io.html#io.BytesIO) has only `tell()` method, not `pos()`:

   https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/4b974204a0d4d811ca5340d1b80acbfcface65b1/kaitaistruct.py#L444-L450

   You have to do just `self` instead of `self._io` to get the `KaitaiStream` object. Note that `self._io` is an instance of `KaitaiStream` only inside a `KaitaiStruct` class.

3. The `src_path` parameter is meant to be a YAML [XPath](https://cs.wikipedia.org/wiki/XPath) inside the source KSY spec pointing to a particular YAML key, not an absolute file path to the `.py` generated source code. The `.py` file path usually isn't much useful information.

   As you don't know the YAML key path here, you shouldn't use any class derived from [`KaitaiStructError`](https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/4b974204a0d4d811ca5340d1b80acbfcface65b1/kaitaistruct.py#L425) at all.

If I were you, I would just raise the standard `Exception` for now in this case. If there's a desire to distinguish between Kaitai Struct errors and the other errors by type, we can think about introducing some exception hierarchy like in https://github.com/kaitai-io/kaitai_struct_php_runtime/pull/6, but please **not** in this PR. It's already quite congested, i.e. not really focusing on one topic.